### PR TITLE
Add 'key' and 'length' methods to Storage

### PR DESCRIPTION
**Pull request by bryophyta**
## What are you changing?

Add wrappers for the built-in `key` method and `length` property on `Storage` objects.

## Why?

These can be useful for extracting the keys from storage, especially given that we (presumably by design?) don't provide direct access to the whole storage object via this API.

Adding `key` and `length` to the API allows users of the library to use these methods with the same extra checks that they get for other methods that wrap built-in functionality in this lib.

In light of the discussion below, I thought it was best to revise this PR to add wrappers around these built-in methods, rather than adding a new abstraction. I thought it best to add these changes on top of the original branch, but happy to rewrite the history to remove the first two commits (which add code which is then removed in the third commit) if desired.

## Original PR description

```
## What are you changing?

Adds a method to return all keys of the storage object in `lib/storage`, by iterating over them.

## Why?

This is not a built-in method on the Storage objects, but provides an abstraction over the built-in `key` method. The `key` method doesn't have that much use outside of getting all the keys, and it has a potential gotcha insofar as the order of keys isn't reliable across user-agents (but `.key()` gets keys by index).

If we aren't happy with adding this new method, could we perhaps consider giving access to the built-in `key` method on the Storage class in this library, to allow users of the library to create their own `keys`-style implementation on top of it?


```

### CODE REVIEW

File: libs/@guardian/libs/src/storage/storage.ts
Comment (mxdvl) [inline libs/@guardian/libs/src/storage/storage.ts:112]: @tjmw reminded me that JavaScript ensure run-to-completion so if the methods are synchronous it's unlikely that such a race condition may occur.

Comment (bryophyta) [reply libs/@guardian/libs/src/storage/storage.ts:112]: This is a good point! Although because the storage objects are browser features I wasn't sure if we could guarantee that the storage side of the operation would be implemented synchronously..

File: libs/@guardian/libs/src/storage/storage.ts
Comment (bryophyta) [reply libs/@guardian/libs/src/storage/storage.ts:131]: Unfortunately storage isn't iterable. As far as I'm away the `key` method is the only supported method for getting keys out of it?

File: libs/@guardian/libs/src/storage/storage.ts
Comment (sndrs) [inline libs/@guardian/libs/src/storage/storage.ts:120]: couple of thoughts:

1. how do you feel about something more explicit, like `allKeys()`? 
2. should we also add `key()` to our api? it only exists to discourage footguns of using the native one. forcing people to use the native one (if they need `key()`) seems wrong

File: libs/@guardian/libs/src/storage/README.md
Comment (sndrs) [inline libs/@guardian/libs/src/storage/README.md:130]: i think it's ok not to mention this. we have other methods that aren't (although the idea of matching the API is intriguing... we just don't already)

File: libs/@guardian/libs/src/storage/README.md
Comment (sndrs) [inline libs/@guardian/libs/src/storage/README.md:131]: > It is therefore potentially subject to race conditions

this makes me wonder if it's something we should have. is it ok to offer a method that might return false information? how is a user supposed to know that? especially if your reading an existing usage that someone else (who did know) wrote?

is it better to some how force ourselves to confront that at the point of doing it, rather than wrapping it in something that _appears_ trustworthy (but can't be)?

File: libs/@guardian/libs/src/storage/README.md
Comment (sndrs) [inline libs/@guardian/libs/src/storage/README.md:139]:
```suggestion
```

Given this describes the implementation, rather than the feature, I'd be inclined to leave this information out of the API description. 

One of the benefits of an abstracted API like this is that you can use it without worrying about what's happening below the surface. You only really need to know that stuff when working on (rather than with) the API itself.

putting this info at the point at which it's used could be very handy, but at this point is maybe distracting rather than informative?

File: libs/@guardian/libs/src/storage/storage.ts
Comment (sndrs) [inline libs/@guardian/libs/src/storage/storage.ts:118]:
```suggestion
```

as with my comment above, it might be useful to move this info to the part of the code where `key` is used. it's certainly useful for people working on the API in the future, but doesn't particularly help people who are using our method (what can they actually do with that info?)

File: libs/@guardian/libs/src/storage/storage.ts
Comment (bryophyta) [reply libs/@guardian/libs/src/storage/storage.ts:122]: I did wonder about this.. I was thinking that it was worth signalling the difference between there not being any keys in the store, and the store not being available. But would be happy to refactor and return `string[]` in all cases if people feel it's worthwhile 👍

File: libs/@guardian/libs/src/storage/README.md
Comment (bryophyta) [reply libs/@guardian/libs/src/storage/README.md:131]: I'll investigate a little more to find out how realistic this issue is.

If we are concerned about this then that would speak in favour of replacing this method with a simple wrapper around the built-in `key` method, as discussed above 👍

File: libs/@guardian/libs/src/storage/storage.ts
Comment (mxdvl) [reply libs/@guardian/libs/src/storage/storage.ts:131]: Just for your information, if you want to avoid mutation you can create an array of a specific length like so:

```ts
return Array.from({ length: this.#storage.length }, (_, i) =>
    this.#storage?.key(i),
    ).filter(isNonNullable);
```

File: libs/@guardian/libs/src/storage/README.md
Comment (bryophyta) [reply libs/@guardian/libs/src/storage/README.md:131]: (partial update here: https://github.com/guardian/csnx/pull/692#discussion_r1261365972)

**Comment by bryophyta** [comment]
@sndrs I've rewritten this to just add wrappers around `key` and `length`, because it seemed from the above that this would be more straightforward than adding a new abstraction. What do you think?

File: libs/@guardian/libs/src/storage/storage.ts
Comment (bryophyta) [inline libs/@guardian/libs/src/storage/storage.ts:121]: I've had the `key` method return `null` if storage is unavailable, because the native method can return `null`, and it seemed best not to introduce a third possible return type. But it generally feels more semantic to return `undefined` if the storage object isn't available so I've used `undefined` for that in the `length` method.

**Comment by jamesmockett** [comment]
Hi @sndrs and @mxdvl 👋  As you've both made suggestions on this PR are you happy if we approve it in its current form?

File: libs/@guardian/libs/src/storage/storage.ts
Comment (mxdvl) [reply libs/@guardian/libs/src/storage/storage.ts:121]: I would actually return `null` or `undefined` everywhere. While I favour `undefined` in general, I think using `null` here would be more consistent.

**Review by mxdvl — APPROVED** [review]
I think this is a worthwhile addition, but would favour a consistent API that returns `null` for all methods when storage is unavailable.

**Review by jamesmockett — APPROVED** [review]
I'll leave it to Max and you to decide whether you want to change the API, but happy to add an approval from CSTI 👍

File: libs/@guardian/libs/src/storage/storage.ts
Comment (bryophyta) [reply libs/@guardian/libs/src/storage/storage.ts:121]: okay thanks, happy to make this change!

**Comment by bryophyta** [comment]
thanks @mxdvl and @jamesmockett ! I've changed the `length` method to return `null` as Max suggested, and tidied up the commit history 👍

**Pull request merged**
